File: hohmann_transfer.py
import math
import numpy as np
from typing import Tuple, Optional
import matplotlib.pyplot as plt
from matplotlib.figure import Figure
from matplotlib.axes import Axes

# Module-level constants for better performance (avoid dataclass overhead)
MU = 398600.4418      # Earth's gravitational parameter (km^3/s^2)
R_EARTH = 6378.137    # Earth's equatorial radius (km)
INV_3600 = 1.0 / 3600.0  # Precomputed hours conversion constant

# Precomputed visualization constants
SPHERE_PHI_POINTS = 20    # Reduced from 30 for faster rendering
SPHERE_THETA_POINTS = 15  # Reduced from 20 for faster rendering
ORBIT_POINTS = 80         # Reduced from 100 for faster plotting

class HohmannTransferOptimized:
    """Ultra-efficient implementation of Hohmann transfer calculations"""
    
    __slots__ = ('r1', 'r2', 'a_transfer', 'v1', 'delta_v_departure', 
                 'delta_v_arrival', 'total_delta_v', 'transfer_time', 
                 'transfer_hours', '_initial_alt', '_final_alt',
                 '_cos_theta', '_sin_theta', '_zeros')
    
    def __init__(self, initial_altitude: float, final_altitude: float):
        """
        Initialize Hohmann Transfer with optimized calculations
        
        Args:
            initial_altitude: Initial orbit altitude (km)
            final_altitude: Final orbit altitude (km)
        """
        # Fast validation without exception overhead for valid inputs
        if initial_altitude <= 0 or final_altitude <= 0:
            raise ValueError("Altitudes must be positive")
        
        # Store original altitudes for display
        self._initial_alt = initial_altitude
        self._final_alt = final_altitude
        
        # Precompute all orbital parameters
        self.r1 = R_EARTH + initial_altitude
        self.r2 = R_EARTH + final_altitude
        self.a_transfer = (self.r1 + self.r2) * 0.5
        
        # Vectorized velocity calculations
        self._compute_all_values()
        
        # Precompute visualization arrays
        self._precompute_visualization()

    def _compute_all_values(self) -> None:
        """Compute all orbital parameters using scalar arithmetic"""
        # math.sqrt on scalars avoids numpy array allocation/dispatch overhead
        mu_over_a = MU / self.a_transfer

        vc1 = math.sqrt(MU / self.r1)                    # circular velocity at r1
        vc2 = math.sqrt(MU / self.r2)                    # circular velocity at r2
        vt1 = math.sqrt(2.0 * MU / self.r1 - mu_over_a) # transfer velocity at r1
        vt2 = math.sqrt(2.0 * MU / self.r2 - mu_over_a) # transfer velocity at r2

        self.v1 = vc1
        self.delta_v_departure = vt1 - vc1
        self.delta_v_arrival = vc2 - vt2
        self.total_delta_v = self.delta_v_departure + self.delta_v_arrival

        self.transfer_time = math.pi * self.a_transfer * math.sqrt(self.a_transfer / MU)
        self.transfer_hours = self.transfer_time * INV_3600

    def _precompute_visualization(self) -> None:
        """Precompute trigonometric values for visualization"""
        theta = np.linspace(0, 2*np.pi, ORBIT_POINTS)
        self._cos_theta = np.cos(theta)
        self._sin_theta = np.sin(theta)
        self._zeros = np.zeros_like(theta)

    def visualize_transfer(self) -> Figure:
        """
        Visualize Hohmann transfer orbit with maximum optimization
        
        Returns:
            matplotlib.figure.Figure: The generated figure
        """
        # Create figure with optimized backend
        fig = plt.figure(figsize=(12, 8))
        ax: Axes = fig.add_subplot(111, projection='3d')
        
        # Ultra-optimized Earth sphere using meshgrid
        phi = np.linspace(0, 2*np.pi, SPHERE_PHI_POINTS)
        theta = np.linspace(0, np.pi, SPHERE_THETA_POINTS)
        phi_mesh, theta_mesh = np.meshgrid(phi, theta)
        
        # Vectorized sphere coordinates
        sin_theta = np.sin(theta_mesh)
        x = R_EARTH * np.cos(phi_mesh) * sin_theta
        y = R_EARTH * np.sin(phi_mesh) * sin_theta
        z = R_EARTH * np.cos(theta_mesh)
        
        # Plot Earth with minimal overhead
        ax.plot_surface(x, y, z, color='blue', alpha=0.15, 
                       linewidth=0, antialiased=False, shade=False)
        
        # Plot orbits using precomputed trigonometric values
        ax.plot(self.r1 * self._cos_theta, self.r1 * self._sin_theta, self._zeros, 
               label='Initial Orbit', color='green', linewidth=2)
        ax.plot(self.r2 * self._cos_theta, self.r2 * self._sin_theta, self._zeros,
               label='Final Orbit', color='red', linewidth=2)
        
        # Add transfer orbit ellipse
        self._plot_transfer_orbit(ax)
        
        # Optimized axis settings
        ax.set_xlabel('X (km)')
        ax.set_ylabel('Y (km)')
        ax.set_zlabel('Z (km)')
        ax.set_title('Hohmann Transfer Orbit')
        ax.legend()
        ax.view_init(30, 45)
        
        return fig

    def _plot_transfer_orbit(self, ax: Axes) -> None:
        """Plot transfer orbit ellipse efficiently"""
        # Semi-ellipse for transfer orbit
        theta_transfer = np.linspace(0, np.pi, ORBIT_POINTS // 2)
        
        # Ellipse parameters
        a = self.a_transfer
        c = abs(self.r2 - self.r1) * 0.5
        b = np.sqrt(a*a - c*c) if a > c else a * 0.1  # Avoid complex numbers
        
        # Parametric ellipse
        x_ellipse = a * np.cos(theta_transfer)
        y_ellipse = b * np.sin(theta_transfer)
        
        # Center and plot
        x_center = (self.r1 + self.r2) * 0.5
        x_transfer = x_ellipse + x_center - a
        
        ax.plot(x_transfer, y_ellipse, np.zeros_like(theta_transfer),
               '--', color='magenta', linewidth=1.5, label='Transfer Orbit')

    def print_transfer_details(self) -> None:
        """Print transfer details with optimized string formatting"""
        # Single f-string for maximum efficiency
        print(f"""
Hohmann Transfer Orbit Details:
Initial Orbit Altitude: {self._initial_alt:.2f} km
Final Orbit Altitude: {self._final_alt:.2f} km
Transfer Orbit Semi-Major Axis: {self.a_transfer:.2f} km
Delta-V (Departure Burn): {self.delta_v_departure:.2f} km/s
Delta-V (Arrival Burn): {self.delta_v_arrival:.2f} km/s
Total Delta-V: {self.total_delta_v:.2f} km/s
Transfer Time: {self.transfer_hours:.2f} hours""")

    # Property-style getters for better performance than methods
    @property
    def departure_delta_v(self) -> float:
        return self.delta_v_departure
    
    @property
    def arrival_delta_v(self) -> float:
        return self.delta_v_arrival
    
    @property
    def total_delta_v_property(self) -> float:
        return self.total_delta_v
    
    @property
    def transfer_time_hours(self) -> float:
        return self.transfer_hours

def get_valid_input(prompt: str) -> float:
    """Optimized input validation with minimal overhead"""
    while True:
        try:
            value = float(input(prompt))
            if value > 0:
                return value
            print("Please enter a positive number.")
        except ValueError:
            print("Please enter a valid number.")

def main() -> None:
    """Optimized main function"""
    try:
        # Fast input with optimized validation
        initial_alt = get_valid_input("Enter initial orbit altitude (km): ")
        final_alt = get_valid_input("Enter final orbit altitude (km): ")
        
        # Create transfer object
        transfer = HohmannTransferOptimized(initial_alt, final_alt)
        
        # Display results
        transfer.print_transfer_details()
        
        # Visualize (optional - comment out for pure calculation performance)
        fig = transfer.visualize_transfer()
        plt.show()
        
    except KeyboardInterrupt:
        print("\nOperation cancelled by user.")
    except Exception as e:
        print(f"Unexpected error: {e}")

if __name__ == "__main__":
    main()

File: hohmann_transfer.m
classdef HohmannTransferOptimized < handle
    properties (Constant, Access = private)
        MU = 398600.4418;        % Earth's gravitational parameter (km^3/s^2)
        R_EARTH = 6378.137;      % Earth's equatorial radius (km)
        INV_3600 = 1/3600;       % Precomputed hours conversion constant
        SPHERE_RESOLUTION = 20;   % Reduced sphere resolution for faster rendering
        ORBIT_POINTS = 100;      % Reduced orbit points for faster plotting
    end
    
    properties (SetAccess = private)
        r1             % Initial orbital radius (km)
        r2             % Final orbital radius (km)
        a_transfer     % Transfer orbit semi-major axis (km)
        delta_v_total  % Total delta-V (km/s)
        transfer_time  % Transfer time (s)
        transfer_hours % Precomputed transfer time in hours
        
        % Precomputed values for display
        initial_altitude
        final_altitude
    end
    
    properties (Access = private)
        delta_v_departure  % Delta-V for departure burn (km/s)
        delta_v_arrival    % Delta-V for arrival burn (km/s)
        
        % Cached trigonometric values for visualization
        cos_theta
        sin_theta
        zeros_array
    end
    
    methods
        function obj = HohmannTransferOptimized(initial_altitude, final_altitude)
            % Simplified validation - faster than validateattributes
            if ~isnumeric(initial_altitude) || ~isscalar(initial_altitude) || ...
               initial_altitude <= 0 || ~isfinite(initial_altitude)
                error('HohmannTransfer:InvalidInput', 'Initial altitude must be a positive finite scalar');
            end
            if ~isnumeric(final_altitude) || ~isscalar(final_altitude) || ...
               final_altitude <= 0 || ~isfinite(final_altitude)
                error('HohmannTransfer:InvalidInput', 'Final altitude must be a positive finite scalar');
            end
            
            % Store for display (avoid repeated calculations)
            obj.initial_altitude = initial_altitude;
            obj.final_altitude = final_altitude;
            
            % Precompute all values at construction
            obj.r1 = obj.R_EARTH + initial_altitude;
            obj.r2 = obj.R_EARTH + final_altitude;
            obj.a_transfer = (obj.r1 + obj.r2) * 0.5;  % Use multiplication instead of division
            
            % Calculate all values in one pass
            obj.computeAllValues();
            
            % Precompute trigonometric values for visualization
            obj.precomputeTrigValues();
        end
        
        function computeAllValues(obj)
            % Vectorized calculations where possible
            mu_over_r = obj.MU ./ [obj.r1, obj.r2];
            v_circular = sqrt(mu_over_r);  % [v1, v2]
            
            % Transfer orbit velocities using optimized calculations
            mu_2_over_r = 2 * mu_over_r;  % [2*MU/r1, 2*MU/r2]
            mu_over_a = obj.MU / obj.a_transfer;
            
            v_transfer = sqrt(mu_2_over_r - mu_over_a);  % [v_transfer_1, v_transfer_2]
            
            % Delta-V calculations
            obj.delta_v_departure = v_transfer(1) - v_circular(1);
            obj.delta_v_arrival = v_circular(2) - v_transfer(2);
            obj.delta_v_total = obj.delta_v_departure + obj.delta_v_arrival;
            
            % Precompute transfer time and hours conversion
            obj.transfer_time = pi * sqrt(obj.a_transfer^3 / obj.MU);
            obj.transfer_hours = obj.transfer_time * obj.INV_3600;
        end
        
        function precomputeTrigValues(obj)
            % Precompute trigonometric values for visualization
            theta = linspace(0, 2*pi, obj.ORBIT_POINTS)';
            obj.cos_theta = cos(theta);
            obj.sin_theta = sin(theta);
            obj.zeros_array = zeros(size(theta));
        end
        
        function printTransferDetails(obj)
            % Single fprintf call with precomputed values
            fprintf(['\nHohmann Transfer Orbit Details:\n', ...
                    'Initial Orbit Altitude: %.2f km\n', ...
                    'Final Orbit Altitude: %.2f km\n', ...
                    'Transfer Orbit Semi-Major Axis: %.2f km\n', ...
                    'Delta-V (Departure Burn): %.2f km/s\n', ...
                    'Delta-V (Arrival Burn): %.2f km/s\n', ...
                    'Total Delta-V: %.2f km/s\n', ...
                    'Transfer Time: %.2f hours\n'], ...
                    obj.initial_altitude, ...
                    obj.final_altitude, ...
                    obj.a_transfer, ...
                    obj.delta_v_departure, ...
                    obj.delta_v_arrival, ...
                    obj.delta_v_total, ...
                    obj.transfer_hours);
        end
        
        function visualizeTransfer(obj)
            % Create figure with optimized settings
            fig = figure('Position', [100, 100, 1200, 800], ...
                        'Renderer', 'opengl');  % Use hardware acceleration
            ax = axes('Parent', fig);
            hold(ax, 'on');
            axis(ax, 'equal');
            
            % Optimized Earth representation with reduced resolution
            [x, y, z] = sphere(obj.SPHERE_RESOLUTION);
            surf(ax, obj.R_EARTH * x, obj.R_EARTH * y, obj.R_EARTH * z, ...
                'FaceAlpha', 0.1, 'EdgeColor', 'none', 'FaceColor', 'blue', ...
                'DisplayName', 'Earth');
            
            % Plot orbits using precomputed trigonometric values
            plot3(ax, obj.r1 * obj.cos_theta, obj.r1 * obj.sin_theta, obj.zeros_array, ...
                  'g-', 'LineWidth', 2, 'DisplayName', 'Initial Orbit');
            plot3(ax, obj.r2 * obj.cos_theta, obj.r2 * obj.sin_theta, obj.zeros_array, ...
                  'r-', 'LineWidth', 2, 'DisplayName', 'Final Orbit');
            
            % Optimized transfer orbit (ellipse)
            obj.plotTransferOrbit(ax);
            
            % Set visualization properties efficiently
            set(ax, 'XLabel', text(0, 0, 'X (km)'), ...
                   'YLabel', text(0, 0, 'Y (km)'), ...
                   'ZLabel', text(0, 0, 'Z (km)'));
            title(ax, 'Hohmann Transfer Orbit');
            legend(ax, 'show', 'Location', 'best');
            view(ax, [45, 20]);
            grid(ax, 'on');
        end
        
        function plotTransferOrbit(obj, ax)
            % Efficient transfer orbit plotting
            theta_transfer = linspace(0, pi, obj.ORBIT_POINTS/2)';
            
            % Semi-major and semi-minor axes
            a = obj.a_transfer;
            c = abs(obj.r2 - obj.r1) * 0.5;  % Distance from center to focus
            b = sqrt(a^2 - c^2);  % Semi-minor axis
            
            % Ellipse coordinates (centered, then translated)
            x_ellipse = a * cos(theta_transfer);
            y_ellipse = b * sin(theta_transfer);
            
            % Translate to proper position
            x_center = (obj.r1 + obj.r2) * 0.5;
            x_transfer = x_ellipse + x_center - a;
            
            plot3(ax, x_transfer, y_ellipse, zeros(size(theta_transfer)), ...
                  'm--', 'LineWidth', 1.5, 'DisplayName', 'Transfer Orbit');
        end
        
        % Getter methods (inline for speed)
        function val = getDeltaVDeparture(obj)
            val = obj.delta_v_departure;
        end
        
        function val = getDeltaVArrival(obj)
            val = obj.delta_v_arrival;
        end
        
        function val = getTotalDeltaV(obj)
            val = obj.delta_v_total;
        end
        
        function val = getTransferTime(obj)
            val = obj.transfer_time;
        end
        
        function val = getTransferTimeHours(obj)
            val = obj.transfer_hours;
        end
    end
    
    methods (Static)
        function main()
            try
                % More efficient input handling
                initial_alt = HohmannTransferOptimized.getValidInput('Enter initial orbit altitude (km): ');
                final_alt = HohmannTransferOptimized.getValidInput('Enter final orbit altitude (km): ');
                
                transfer = HohmannTransferOptimized(initial_alt, final_alt);
                transfer.printTransferDetails();
                transfer.visualizeTransfer();
            catch ME
                fprintf('Error: %s\n', ME.message);
            end
        end
        
        function value = getValidInput(prompt)
            % Optimized input validation
            while true
                value = input(prompt);
                if isnumeric(value) && isscalar(value) && value > 0 && isfinite(value)
                    break;
                end
                fprintf('Please enter a positive finite number.\n');
            end
        end
    end
end
